### ISSUE TITLE
More Parallelization

### ISSUE THREAD

**Issue by troydalldorf** [graphql-dotnet/graphql-dotnet#889]
From what I can tell looking at Fiddler there seems to be room for improving the parallel execution. The current model seems to run one nodes children in parallel, but then wait for all of those to complete before continuing down the graph.

I would like to see each completed node immediately start work on the child nodes.

ParallelExecutionStrategy does indeed wait for each set of child nodes.

**Comment by troydalldorf** [comment]
A nice-to-have addition to this would be to have the ability to process children of children in parallel where the parent id was the same or said differently where the child and the child of child have a 1:1 ratio. For example, if an organization only has one address, it would be great to have organization and its address child loaded in parallel.

**Comment by troydalldorf** [comment]
The nice-to-have in mu comment is already available (#354)

**Comment by troydalldorf** [comment]
@joemcbride I have a fork with the code [here](https://github.com/troydalldorf/graphql-dotnet/commit/83326b91d491f7ddb40d5a42ea765fe8530a6423).
It's working code, but pretty rough, i.e. no comments, testing, or ability to set settings, propagate task cancellation, etc. and I wasn't sure how to handle: OnBeforeExecutionStepAwaitedAsync(context).
If this is something that would be considered being merged into graphql-dotnet master, I can clean it up and add testing.
As a side note: Are there any plans to allow dependency injection of the ExecutionStrategy?

**Comment by johnrutherford** [comment]
I wrote the ParallelExecutionStrategy, so I’d certainly be interested in checking this out in a pull request. 

I’d be a little concerned if there would be some scaling issues if a new Dataflow pipeline is created for each execution. There could also be issues with running in a SynchronizationContext (like with #795). I’m also not sure how it would work batching queries with the DataLoader. 

At the very least, maybe it could be available as an alternate execution strategy that users could opt-in to.

**Comment by johnrutherford** [comment]
@troydalldorf, I’ll check it out. The DataLoader requires `OnBeforeExecutionStepAwaitedAsync` to be called or it will deadlock. 

Currently one can sub-class `DocumentExecuter` and override `SelectExecutionStrategy()` to use a different execution strategy.

**Comment by troydalldorf** [comment]
@johnrutherford, here's the pull request:
https://github.com/graphql-dotnet/graphql-dotnet/pull/890

Still has plenty of work to do, but would appreciate the feedback. Thanks!

**Comment by johnrutherford** [comment]
@troydalldorf, you inspired me to look into how the existing `ParallelExecutionStrategy` could be optimized to allow sub-nodes to execute once parent nodes are complete. It hadn't occurred to me at the time when I first wrote it. 

I opened pull request ###. Can you check it out to see if it meets your needs?

There could still be a use case for an execution strategy using DataFlow blocks, so I'll check out your PR when I have time tomorrow.

**Comment by troydalldorf** [comment]
Thanks. I'll check out your pull request as well. Thanks for digging into this so fast!

**Comment by sungam3r** [comment]
Closing, see https://github.com/graphql-dotnet/graphql-dotnet/pull/890#issuecomment-786302639
